File: 141_very_hard_Matrix_Multiplication_Part_2.js
function matrix_multiply(a, b){

    if (a[0].length != b.length){
        return "invalid";
    }

    // Initialzie the result matrix with zeros
    let result = new Array(a.length);
    for (let i = 0; i < a.length; i++){
        result[i] = new Array(b[0].length).fill(0);
    }

    // Perform matrix multiplication
    for(let i = 0; i < a.length; i++){
        for (let j = 0; j < b[0].length; j++){
            for (let k = 0; k < b.length; k++){
                result[i][j] += a[i][k] * b[k][j];
            }
        }
    }

    return result;
}

// Test cases
console.log(matrix_multiply([[1, 2]], [[3], [4]])); // ➞ [[11]]
console.log(matrix_multiply([[0, 0], [0, 1]], [[1, 2], [3, 4], [5, 6]])); // ➞ "invalid"
console.log(matrix_multiply([[4, 2], [3, 1]], [[5, 6], [3, 8]])); // ➞ [[26, 40], [18, 26]]

File: 141_very_hard_Matrix_Multiplication_Part_2.py
"""
Matrix Multiplication (Part 2)
Create a function that multiplies two matrices (n x m) and (p x q) and returns:

"invalid" if the matrices are not multiplicable (i.e. if m is not equal to p).
The multiplication matrix (n x q) otherwise.
Examples
matrix_multiply([[1, 2]], [[3], [4]]) ➞ [[11]]

matrix_multiply([[0, 0], [0, 1]], [[1, 2], [3, 4], [5, 6]]) ➞ "invalid"

matrix_multiply([[4, 2], [3, 1]], [[5, 6], [3, 8]]) ➞ [[26, 40], [18, 26]]
"""

def matrix_multiply(a, b):

    if len(a[0]) != len(b):
        return "invalid"

    result = [[0] * len(b[0]) for _ in range(len(a))]

    for i in range(len(a)):
        for j in range(len(b[0])):
            for k in range(len(b)):
                result[i][j] += a[i][k] * b[k][j]

    return result

print(matrix_multiply([[0, 0], [0, 1]], [[1, 2], [3, 4], [5, 6]]))
print(matrix_multiply([[4, 2], [3, 1]], [[5, 6], [3, 8]]))
print(matrix_multiply([[1, 6], [6, 3]], [[5, 3, 5], [1, 6, 6]]))
